lib/src/value/number.ts

```diff
@@ -0,0 +1,718 @@
+      newDenominatorUnits: asImmutableList(newDenominators),
+      name,
+    });
+  }
+
+  /**
+   * Returns a copy of `this`, converted to the same units as `other`.
+   *
+   * Does *not* throw an error if `this` is unitless and `other` is not, or
+   * vice-versa. Instead, it treats all unitless numbers as convertible to and
+   * from all units without changing the value.
+   *
+   * Throws an error if `this`'s units are incompatible with `other`'s units.
+   *
+   * If `this` came from a function argument, `name` is the argument name
+   * and `otherName` is the argument name for `other` (both without the `$`).
+   * They are used for error reporting.
+   */
+  coerceToMatch(
+    other: SassNumber,
+    name?: string,
+    otherName?: string
+  ): SassNumber {
+    return SassNumber.withUnits(
+      this.coerceValueToMatch(other, name, otherName),
+      {
+        numeratorUnits: other.numeratorUnits,
+        denominatorUnits: other.denominatorUnits,
+      }
+    );
+  }
+
+  /**
+   * Returns `value`, converted to the same units as `other`.
+   *
+   * Does *not* throw an error if `this` is unitless and `other` is not, or
+   * vice-versa. Instead, it treats all unitless numbers as convertible to and
+   * from all units without changing the value.
+   *
+   * Throws an error if `this`'s units are incompatible with `other`'s units.
+   *
+   * If `this` came from a function argument, `name` is the argument name
+   * and `otherName` is the argument name for `other` (both without the `$`).
+   * They are used for error reporting.
+   */
+  coerceValueToMatch(
+    other: SassNumber,
+    name?: string,
+    otherName?: string
+  ): number {
+    return convertOrCoerce(this, {
+      coerceUnitless: true,
+      other,
+      name,
+      otherName,
+    });
+  }
+
+  equals(other: Value): boolean {
+    if (!(other instanceof SassNumber)) return false;
+    try {
+      return fuzzyEquals(this.value, other.convertValueToMatch(this));
+    } catch {
+      return false;
+    }
+  }
+
+  hashCode(): number {
+    const canonicalNumerators = canonicalizeUnits(this.numeratorUnits);
+    const canonicalDenominators = canonicalizeUnits(this.denominatorUnits);
+    const canonicalValue = this.convertValue(
+      canonicalNumerators,
+      canonicalDenominators
+    );
+    return (
+      fuzzyHashCode(canonicalValue) ^
+      hash(canonicalNumerators) ^
+      hash(canonicalDenominators)
+    );
+  }
+
+  toString(): string {
+    return 'SassNumber';
+  }
+}
+
+// Returns the value of converting `number` to new units.
+//
+// The units may be specified as lists of units (`newNumeratorUnits` and
+// `newDenominatorUnits`), or by providng a SassNumber `other` that contains the
+// desired units.
+//
+// Throws an error if `number` is not compatible with the new units. Coercing a
+// unitful number to unitless (or vice-versa) throws an error unless
+// specifically enabled with `coerceUnitless`.
+function convertOrCoerce(
```

Comment:
If you're always passing `this` as the first argument, why not just make this a private instance method?